librad/src/identities/git.rs

```diff
@@ -0,0 +1,760 @@
+
+#[derive(Clone)]
+pub struct Git<'a, T> {
+    repo: &'a git2::Repository,
+    _marker: PhantomData<T>,
+}
+
+impl<'a, T: 'a> Git<'a, T> {
+    pub fn new(repo: &'a git2::Repository) -> Self {
+        Self {
+            repo,
+            _marker: PhantomData,
+        }
+    }
+
+    /// Convenience to specialise `T` to [`User`].
+    pub fn as_user(&self) -> Git<'_, User> {
+        Git {
+            repo: self.repo,
+            _marker: PhantomData,
+        }
+    }
+
+    /// Convenience to specialise `T` to [`Project`].
+    pub fn as_project(&self) -> Git<'_, Project> {
+        Git {
+            repo: self.repo,
+            _marker: PhantomData,
+        }
+    }
+
+    /// Read an identity whose type is not statically known from commit `oid`.
+    ///
+    /// The only guarantee about the returned value is that it is well-formed --
+    /// it may or may not pass verification.
+    pub fn some_identity(&self, oid: git2::Oid) -> Result<SomeIdentity, error::Load> {
+        SomeIdentity::try_from(self.by_oid(oid))
+    }
+
+    /// Traverse the history with head commit `head`, yielding identities of
+    /// type `T`.
+    ///
+    /// The iterator yields elements in **reverse order**, ie. oldest-first. No
+    /// verification is performed, but `Ok` elements are guaranteed to be
+    /// well-formed.
+    pub fn iter(
+        &self,
+        head: git2::Oid,
+    ) -> Result<impl Iterator<Item = Result<T, error::Load>> + 'a, error::Load>
+    where
+        T: TryFrom<ByOid<'a>, Error = error::Load>,
+    {
+        Ok(Iter::new(self.repo, head)?
+            .map(|item: Result<generic::Verifying<T, _>, _>| item.map(|v| v.into_inner())))
+    }
+
+    //// Generic methods ////
+
+    pub fn get_generic(&self, oid: git2::Oid) -> Result<T, error::Load>
+    where
+        T: TryFrom<ByOid<'a>, Error = error::Load>,
+    {
+        T::try_from(self.by_oid(oid))
+    }
+
+    fn verify_generic<Doc>(
+        &self,
+        head: git2::Oid,
+    ) -> Result<VerifiedIdentity<Doc>, VerificationError>
+    where
+        Doc: Delegations + generic::Replaces<Revision = Revision>,
+        <Doc as Delegations>::Error: std::error::Error + Send + Sync + 'static,
+
+        Identity<Doc>: TryFrom<ByOid<'a>, Error = error::Load>,
+    {
+        self.fold_verify_generic(head).map(|folded| folded.head)
+    }
+
+    fn fold_verify_generic<Doc>(
+        &self,
+        head: git2::Oid,
+    ) -> Result<generic::Folded<Doc, Revision, ContentId>, VerificationError>
+    where
+        Doc: Delegations + generic::Replaces<Revision = Revision>,
+        <Doc as Delegations>::Error: std::error::Error + Send + Sync + 'static,
+
+        Identity<Doc>: TryFrom<ByOid<'a>, Error = error::Load>,
+    {
+        let mut progeny = Iter::<'_, Identity<Doc>>::new(self.repo, head)
+            .map_err(generic::error::Verify::history)?;
+
+        // TODO(kim): should we skip non-quorum commits at the beginning?
```

Comment:
It is actually the case that this doesn't _have_ to start at the first commit. It would be perfectly possible to create the initial revision with any number > 1 of trust delegations, but sign that only with the key at hand (which means no quorum). There isn't a way, however, to collect the signatures until quorum is reached _in band_, because replication demands to find a verified revision (that's the whole point).

So, the only reasons to skip until we find a quorum would be to support some kind of oob initialisation -- for example, via your favourite hosting service -- or to propose a new identity within the context of an already-valid project (and later just transplanting that branch). Idk if it is preferable to adhere to Postel's Law here, or to encode the youre-supposed-to-hold-it-this-way path and fix things once we learn about what crazy things people are trying to do.